integration_tests/src/main/scala/org/apache/spark/sql/tests/datagen/big_data_gen.scala

```diff
@@ -0,0 +1,1923 @@
+  require(values != null && values.length > 0)
+
+  override def withLocationToSeedMapping(mapping: LocationToSeedMapping): EnumValGenFunc = {
+    EnumValGenFunc(values, mapping)
+  }
+
+  override def withValueRange(min: Any, max: Any): GeneratorFunction =
+    throw new IllegalArgumentException(s"value ranges are not supported for EnumValueGenFunc")
+
+  override def apply(rowLoc: RowLocation): Any = {
+    val r = DataGen.getRandomFor(rowLoc, mapping)
+    values(r.nextInt(values.length))
+  }
+}
+
+/**
+ * A value generator for booleans.
+ */
+case class BooleanGenFunc(mapping: LocationToSeedMapping = null) extends GeneratorFunction {
+  override def withLocationToSeedMapping(mapping: LocationToSeedMapping): BooleanGenFunc =
+    BooleanGenFunc(mapping)
+
+  override def apply(rowLoc: RowLocation): Any =
+    DataGen.getRandomFor(rowLoc, mapping).nextBoolean()
+
+  override def withValueRange(min: Any, max: Any): GeneratorFunction = {
+    val minb = BooleanGen.asBoolean(min)
+    val maxb = BooleanGen.asBoolean(max)
+    if (minb == maxb) {
+      SingleValGenFunc(minb)
+    } else {
+      this
+    }
+  }
+}
+
+object BooleanGen {
+  def asBoolean(a: Any): Boolean = a match {
+    case b: Boolean => b
+    case other =>
+      throw new IllegalArgumentException(s"a boolean value range only supports " +
+          s"boolean values found $other")
+  }
+}
+
+class BooleanGen(conf: ColumnConf,
+    defaultValueRange: Option[(Any, Any)])
+    extends DataGen(conf, defaultValueRange) {
+
+  override def dataType: DataType = BooleanType
+
+  override protected def getValGen: GeneratorFunction = BooleanGenFunc()
+}
+
+/**
+ * A value generator for Bytes
+ */
+case class ByteGenFunc(mapping: LocationToSeedMapping = null,
+    min: Byte = Byte.MinValue,
+    max: Byte = Byte.MaxValue) extends GeneratorFunction {
+
+  private lazy val valueRemapping: Long => Byte = ByteGen.remapRangeFunc(min, max)
+
+  override def apply(rowLoc: RowLocation): Any = valueRemapping(DataGen.nextLong(rowLoc, mapping))
+
+  override def withLocationToSeedMapping(mapping: LocationToSeedMapping): ByteGenFunc =
+    ByteGenFunc(mapping, min, max)
+
+  override def withValueRange(min: Any, max: Any): GeneratorFunction = {
+    var bmin = ByteGen.asByte(min)
+    var bmax = ByteGen.asByte(max)
+    if (bmin > bmax) {
+      val tmp = bmin
+      bmin = bmax
+      bmax = tmp
+    }
+
+    if (bmin == bmax) {
+      SingleValGenFunc(bmin)
+    } else {
+      ByteGenFunc(mapping, bmin, bmax)
+    }
+  }
+}
+
+object ByteGen {
+  def asByte(a: Any): Byte = a match {
+    case n: Byte => n
+    case n: Short if n <= Byte.MaxValue && n >= Byte.MinValue =>
+      n.toByte
+    case n: Int if n <= Byte.MaxValue && n >= Byte.MinValue =>
+      n.toByte
+    case n: Long if n <= Byte.MaxValue && n >= Byte.MinValue =>
+      n.toByte
+    case other =>
+      throw new IllegalArgumentException(s"a byte value range only supports " +
+          s"byte values found $other")
+  }
```

Comment:
If the intent is to allow type conversions as long as the original value is preserved in the new form, we could simplify this to:
```suggestion
  def asByte(a: Any): Byte = {
    val converted = a.toByte
    if (converted == a) {
      converted
    } else {
      throw new IllegalArgumentException(s"a byte value range only supports " +
          s"byte values found $other")
    }
  }
```
Which would then allow byte conversions for floating point, decimals, etc as long as their value was completely convertible into a byte (which seems to meet the spirit of what was coded for short, int, and long).

Reply:
`a` is an Any. There is no `toByte` for `a`. a Scala Byte class only inherits from AnyVal, so there is not even a type hierarchy to integer values/etc. I'll try to make it work with java's Number. It looks like scala plays nicely with that.